### PR TITLE
[Core] Allow users to specify the classpath and import path

### PR DESCRIPTION
**Pull request by chaokunyang**
## Why are these changes needed?
Allow users to specify the classpath of Java workers and import path of python workers in job config or in ray start options

Now user canset code serach path like this:
## Start a new cluster
```bash
ray start --head --redis-port=6379 --redis-password=123456 --include-java --code-search-path="$PWD/bazel-bin/java/all_tests_deploy.jar"
```
## Connecting to exiting cluster
```bash
java -cp bazel-bin/java/all_tests_deploy.jar -Dray.redis.address='10.15.239.35:6379' -Dray.redis.password='5241590000000000' -Dray.job.resource-path=$PWD/bazel-bin/java/ io.ray.test.MultiDriverTest
```
## Related issue number
 #10433

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by kfstorm** [comment]
Can we add a unit test to cover this?

**Comment by kfstorm** [comment]
Please support configuration this option in Python.

File: java/runtime/src/test/java/io/ray/runtime/functionmanager/FunctionManagerTest.java [hunk 1]
```diff
@@ -151,7 +153,8 @@ public void testLoadFunctionTableForClass() {
   @Test
   public void testGetFunctionFromLocalResource() throws Exception {
     JobId jobId = JobId.fromInt(1);
-    final String resourcePath = FileUtils.getTempDirectoryPath() + "/ray_test_resources";
+    final List<String> resourcePath =
+        Collections.singletonList(FileUtils.getTempDirectoryPath() + "/ray_test_resources");
     final String jobResourcePath = resourcePath + "/" + jobId.toString();
```
[the remark this answers is not in the document]
Comment (chaokunyang) [reply java/runtime/src/test/java/io/ray/runtime/functionmanager/FunctionManagerTest.java]: fixed

File: java/runtime/src/test/java/io/ray/runtime/functionmanager/FunctionManagerTest.java [hunk 2]
```diff
@@ -151,8 +152,7 @@ public void testLoadFunctionTableForClass() {
   @Test
   public void testGetFunctionFromLocalResource() throws Exception {
     JobId jobId = JobId.fromInt(1);
```
Comment (kfstorm) [inline java/runtime/src/test/java/io/ray/runtime/functionmanager/FunctionManagerTest.java:154]: Unused variable?

File: java/test.sh [hunk 3]
```diff
@@ -43,6 +43,15 @@ echo "Running tests under single-process mode."
...
+echo "Running connecting existing cluster tests"
+case "${OSTYPE}" in
+  linux*) ip=$(hostname -I | awk '{print $1}');;
+  darwin*) ip=$(ipconfig getifaddr en0);;
+  *) echo "Can't get ip address for ${OSTYPE}"; exit 1;;
+esac
+RAY_BACKEND_LOG_LEVEL=debug ray start --head --redis-port=6379 --redis-password=123456 --include-java --job-resource-path=[\""$PWD/bazel-bin/java/all_tests_deploy.jar"\"]
+RAY_BACKEND_LOG_LEVEL=debug java -cp bazel-bin/java/all_tests_deploy.jar -Dray.redis.address="$ip:6379"\
+ -Dray.redis.password='123456' -Dray.job.resource-path.0="$PWD/bazel-bin/java/all_tests_deploy.jar" io.ray.test.MultiDriverTest
 popd
```
Comment (kfstorm) [inline java/test.sh:58]: Add `ray stop` here

File: src/ray/protobuf/gcs.proto [hunk 4]
```diff
@@ -321,6 +321,8 @@ message JobConfig {
   uint32 num_java_workers_per_process = 2;
   // The jvm options for java workers of the job.
   repeated string jvm_options = 3;
+  // job resource directory
```
Comment (kfstorm) [inline src/ray/protobuf/gcs.proto:324]:
```suggestion
  // Job resource directories
```

File: src/ray/protobuf/gcs.proto [hunk 5]
```diff
@@ -321,6 +321,8 @@ message JobConfig {
...
+  repeated string job_resource_path = 4;
```
Comment (kfstorm) [inline src/ray/protobuf/gcs.proto:325]:
```suggestion
  repeated string resource_path = 4;
```

File: java/runtime/src/test/java/io/ray/runtime/functionmanager/FunctionManagerTest.java (hunk 2 above)
Comment (chaokunyang) [reply java/runtime/src/test/java/io/ray/runtime/functionmanager/FunctionManagerTest.java:154]: `io.ray.runtime.functionmanager.FunctionManager#getFunction(io.ray.api.id.JobId, io.ray.runtime.functionmanager.JavaFunctionDescriptor)` need a job id

File: python/ray/worker.py [hunk 6]
```diff
@@ -1216,6 +1216,8 @@ def connect(node,
         node.node_manager_port, node.raylet_ip_address, (mode == LOCAL_MODE),
         driver_name, log_stdout_file_path, log_stderr_file_path,
         serialized_job_config, node.metrics_agent_port)
+    job_resource_path = job_config.job_resource_path
```
[the remark this answers is not in the document]
Comment (chaokunyang) [reply python/ray/worker.py]: removed

File: python/ray/services.py [hunk 7]
```diff
@@ -1438,9 +1442,12 @@ def start_raylet(redis_address,
...
         f"--config_list={config_str}",
-        f"--python_worker_command={subprocess.list2cmdline(start_worker_command)}",  # noqa
-        f"--java_worker_command={subprocess.list2cmdline(java_worker_command)}",  # noqa
-        f"--cpp_worker_command={subprocess.list2cmdline(cpp_worker_command)}",  # noqa
+        f"--python_worker_command={subprocess.list2cmdline(start_worker_command)}",
+        # noqa
+        f"--java_worker_command={subprocess.list2cmdline(java_worker_command)}",
+        # noqa
+        f"--cpp_worker_command={subprocess.list2cmdline(cpp_worker_command)}",
+        # noqa
```
[the remark this answers is not in the document]
Comment (chaokunyang) [reply python/ray/services.py]: reverted

File: java/test.sh (hunk 3 above)
Comment (chaokunyang) [reply java/test.sh:58]: Good catch

File: src/ray/protobuf/gcs.proto (hunk 5 above)
Comment (chaokunyang) [reply src/ray/protobuf/gcs.proto]: renamed

File: java/runtime/src/main/resources/ray.default.conf [hunk 8]
```diff
@@ -22,10 +22,11 @@ ray {
...
     id: ""
-    // If this config is set, worker will use different paths to load resources when
-    // executing tasks from different jobs. E.g. if it's set to '/tm/job_resources',
-    // the path for job 123 will be '/tmp/job_resources/123'.
-    resource-path: ""
+    // If this config is set, worker will use different paths to load resources
+    // when executing tasks from different jobs. E.g. if it's set to
+    // '/tmp/job_resources,/tmp/job_resources2', the path for job 123 will be
+    // ['/tmp/job_resources', '/tmp/job_resources2'].
+    resource-path: []
```
[the remark this answers is not in the document]
Comment (chaokunyang) [reply java/runtime/src/main/resources/ray.default.conf:27]: Like classpath, `resource-path` can represent multiple directories?

File: python/ray/job_config.py [hunk 9]
```diff
@@ -10,13 +10,17 @@ class JobConfig:
         num_java_workers_per_process (int): The number of java workers per
             worker process.
         jvm_options (str[]): The jvm options for java workers of the job.
+        code_search_path (str): A list of directories or jar files that specify
```
Comment (kfstorm) [inline python/ray/job_config.py:13]:
```suggestion
        code_search_path (list): A list of directories or jar files that specify
```

**Review by raulchen — APPROVED** [review]

**Pull request merged**